### ISSUE TITLE
qt5-qtwebengine for aarch64 and armhf isn't in the binary repo

### ISSUE THREAD

**Issue by ollieparanoid** [postmarketOS/pmbootstrap#1008]
`qt5-qtwebengine` is the whole `chromium` browser, with a bit of `QT` glue. It takes hours to compile, even with a fast CPU. I've tried to compile it for `x86_64`, `armhf` and `aarch64`.

### Status
* `x86_64`: compiled, and is available in the repo now (I did not test using it, that would be helpful)
* `armhf` failed at the very end when linking `libQt5WebEngineCore`
```
linking ..../lib/libQt5WebEngineCore.so.5.9.3
collect2: fatal error: ld terminated with signal 11 [Segmentation fault], core dumped
compilation terminated
distcc[12345] ERROR: compile (null) on localhost failed
```
* `aarch64`: hangs forever (left it there for hours) on `[327/237] LINK gn` fairly early for some reason.

**NOTE:** I gave it 50 GB of SWAP just in case, so running out of memory shouldn't be the problem (and that would have gotten other output).

### How to help
Try to compile it on your own PC for foreign arches. Especially for `aarch64` it fails early for me, so it would be good to know if it's just me or not. If it's possible, trying to compile on a native device (e.g. @opendata26 with his Shield TV) would also be helpful (does it run through there?).

To build it, change `arch="x86_64"` back to `arch="all"` in `aports/main/qt5-qtwebengine/APKBUILD`, then run either:
* `pmbootstrap build --arch=aarch64 qt5-qtwebengine`
* `pmbootstrap build --arch=armhf qt5-qtwebengine`

**Comment by ollieparanoid** [comment]
I've tried to build for `aarch64` again, and it hangs again. But this time it also printed the following (I guess I missed that somehow last time?):
```
[327/327] LINK gn
terminate called after throwing an instance of `std::bad_alloc`
terminate called recursively
Command '['/home/pmos/build/src/buil/src/3rdparty/chromium/tools/gn/out/Release/gn', 'gen', ...]' returnet non-zero exit status -6
```

**Comment by z3ntu** [comment]
I've tried building it natively on my Raspberry Pi 3 with ArchLinuxARM but cc1plus just gets killed after some time. (`g++: internal compiler error: Killed (program cc1plus)`). Probably RAM related or something.
I've made a 4GB swap file and am trying again but I don't expect it to work...

**Comment by z3ntu** [comment]
Result over the night: Xorg got killed while running but it got to about 8000/18000 instead of the only 4000/18000 prior to activating swap. So natively compiling is not an option on the RPi3 😉 (at least for qtwebengine)

**Comment by ollieparanoid** [comment]
Thanks for trying :+1:

I have another idea: Cross-compiling from the native chroot (just like we do it for the kernel). That would require the target arch buildroot to be mounted inside the native chroot (similar to #665). All in all I think it's quite some work, but since compiling *does* work for `x86_64`, I guess cross-compiling would work then as well. [See](https://wiki.linaro.org/WorkingGroups/Middleware/Graphics/Qt/CrossCompiling) [also](https://blog.qt.io/blog/2012/04/13/cross-compiling-qt-for-the-masses/).

**Comment by z3ntu** [comment]
And I can reproduce the armhf issue with cross-compilation from x86_64, so it's not because of your computer 😉 
```
ar cqs Release/libqtwebenginecoreapi.a .obj/qtwebenginecoreglobal.o .obj/qwebenginecookiestore.o .obj/qwebenginehttprequest.o .obj/qwebengineurlrequestinfo.o .obj/qwebengineurlrequestjob.o .obj/qwebengineurlschemehandler.o .obj/qtbug-60565.o .obj/qtbug-61521.o .obj/moc_qwebenginecookiestore.o .obj/moc_qwebengineurlrequestinterceptor.o .obj/moc_qwebengineurlrequestjob.o .obj/moc_qwebengineurlschemehandler.o
make[3]: Leaving directory '/home/pmos/build/src/build/src/core/api'
( test -e Makefile.core_module || /usr/bin/qmake-qt5 -o Makefile.core_module /home/pmos/build/src/qtwebengine-opensource-src-5.9.3/src/core/core_module.pro WEBENGINE_CONFIG+=use_system_ffmpeg ) && /usr/bin/make -f Makefile.core_module 
make[3]: Entering directory '/home/pmos/build/src/build/src/core'
creating linker version script QtWebEngineCore.version
rm -f libQt5WebEngineCore.so.5.9.3 libQt5WebEngineCore.so libQt5WebEngineCore.so.5 libQt5WebEngineCore.so.5.9
linking ../../lib/libQt5WebEngineCore.so.5.9.3
collect2: fatal error: ld terminated with signal 11 [Segmentation fault], core dumped
compilation terminated.
distcc[3560] ERROR: compile (null) on localhost failed
make[3]: *** [Makefile.core_module:77: ../../lib/libQt5WebEngineCore.so.5.9.3] Error 1
make[3]: Leaving directory '/home/pmos/build/src/build/src/core'
make[2]: *** [Makefile:122: sub-core_module-pro-make_first] Error 2
make[2]: Leaving directory '/home/pmos/build/src/build/src/core'
make[1]: *** [Makefile:79: sub-core-make_first] Error 2
make[1]: Leaving directory '/home/pmos/build/src/build/src'
make: *** [Makefile:47: sub-src-make_first] Error 2
>>> ERROR: qt5-qtwebengine: all failed
```

**Comment by magmastonealex** [comment]
I was also able to reproduce both results, unfortunately. 

@z3ntu, I noticed ld dumped core, could you try to examine the coredump with gdb to maybe see what the cause of the segfault was? Hopefully it will be something easier to fix than cross-compiling qtwebengine. I would check mine, but I ran a `zap` before noticing the coredump :(

**Comment by z3ntu** [comment]
```
/home/pmos/build/src/build/src/core # /usr/lib/gcc/armv6-alpine-linux-musleabihf/6.4.0/../../../../armv6-alpine-linux-musleabihf/bin/ld @/tmp/cchaahmL
Segmentation fault (core dumped)
/home/pmos/build/src/build/src/core # gdb --args /usr/lib/gcc/armv6-alpine-linux-musleabihf/6.4.0/../../../../armv6-alpine-linux-musleabihf/bin/ld @/tmp/cchaahmL
GNU gdb (GDB) 8.0.1
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "armv6-alpine-linux-musleabihf".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/lib/gcc/armv6-alpine-linux-musleabihf/6.4.0/../../../../armv6-alpine-linux-musleabihf/bin/ld...(no debugging symbols found)...done.
(gdb) run
Starting program: /usr/armv6-alpine-linux-musleabihf/bin/ld @/tmp/cchaahmL
qemu: Unsupported syscall: 26
Could not trace the inferior process.
Error: ptrace: Function not implementedDuring startup program exited with code 127.
```
Apparently you can't use gdb directly inside a qemu chroot but you need to use gdbserver or something (see https://tinkering-is-fun.blogspot.co.at/2009/12/debugging-non-native-programs-with-qemu.html)

**Comment by magmastonealex** [comment]
@z3ntu I think having the coredump already will mean you don't need to run the program within gdb, but you can still analyze what happened.

What happens if you run:

* `ulimit -c unlimited` to enable coredump file output
* The ld command you created to reproduce this
* `gdb /path/to/ld /path/to/coredump/core` 

The coredump should be in the directory you ran everything from, in a file called `core`. I believe that contains enough information for gdb to recreate the state of the program without needing ptrace or other hardware debug features.

**Comment by z3ntu** [comment]
(because I use systemd and systemd likes systemd-coredump, you have to change it:
```
cat /proc/sys/kernel/core_pattern # in chroot
sudo sysctl -w kernel.core_pattern=/var/crash/core.%u.%e.%p # on host
cat /proc/sys/kernel/core_pattern # in chroot
```
then
```
/home/pmos/build/src/build/src/core # gdb /usr/lib/gcc/armv6-alpine-linux-musleabihf/6.4.0/../../../../armv6-alpine-linux-musleabihf/bin/ld /var/crash/core.0.ld.28721
GNU gdb (GDB) 8.0.1
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "armv6-alpine-linux-musleabihf".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/lib/gcc/armv6-alpine-linux-musleabihf/6.4.0/../../../../armv6-alpine-linux-musleabihf/bin/ld...(no debugging symbols found)...done.
"/var/crash/core.0.ld.28721" is not a core dump: File format not recognized
(gdb) 
```
any idea? the file contains binary stuff

**Comment by ollieparanoid** [comment]
Thanks for confirming that it's not my computer 😃 

I have looked into cross-compiling QT from the native chroot. The big problem is, that we would need to call QT's [configure](https://doc.qt.io/qt-5/configure-options.html) to prepare everything for cross-compilation. And the configure script is only shipped in the full QT source, but not in each module's source code. This means one would need to compile all of QT in one go, and not the way Alpine does it, by compiling each module in a single package (which makes more sense from a packager's point of view).

As alternative it might be possible to modify some `*.pro` files in the source tree, [like done here](https://github.com/postmarketOS/pmbootstrap/issues/878#issuecomment-352292328)?

Anyway, I put my WIP changes in the `feature/native-cross-compile-qt` branch if someone wants to play with it (based on @zhuowei's work in #665, see the commit description of e0f90e2f8df5f5e8140114f7aff67140cbf0182b for details).

**Comment by magmastonealex** [comment]
I was doing some digging trying to figure out why this might be happening, and stumbled upon [this bug](https://bugs.launchpad.net/ubuntu/+source/qemu-linaro/+bug/906922).

It's been resolved, but it pointed out that when qemu is doing 32bit binfmt emulation, the emulated utilities won't be able to use more than 2G of ram - which makes sense, even though it's a 64 bit system, the utilities are still 32bit.

That may be the cause of the segfault - I bet that final `ld` call will use a lot more than 2G of ram, and a swapfile won't help.

**Comment by zhuowei** [comment]
@ollieparanoid if you need to check the correct command line arguments to pass to qmake for cross compiling, try looking at OpenEmbedded's qtwebengine recipe (which LuneOS uses). https://github.com/meta-qt5/meta-qt5/blob/master/recipes-qt/qt5/qtwebengine_git.bb and the generic qmake5 base recipe (which the qtwebchannel recipe uses). https://github.com/meta-qt5/meta-qt5/blob/master/classes/qmake5_base.bbclass

Seems they just write a temp qt.conf into the local directory directly when building instead of running qt's config script to set up paths.

**Comment by z3ntu** [comment]
Interesting fact:
ArchLinuxARM has `qt5-base` for `arm` (v5 I think), `armv6h`, `armv7h` and `aarch64` but `qt5-webengine` (equivalent to `qt5-qtwebengine` here) only for `armv7h` and `aarch64`...

**Comment by z3ntu** [comment]
Also in ArchLinuxARM chromium is only available for armv7h and aarch64 and on Alpine it's only available for x86_64.

**Comment by JBBgameich** [comment]
If I'm not mistaken, qtwebengine is available for armhf in Debian. Maybe you can copy a few more patches from there: https://anonscm.debian.org/git/pkg-kde/qt/qtwebengine.git

**Comment by z3ntu** [comment]
@JBBgameich I didn't find anything about arm* in the debian patches but the rules file could be interesting for us. They specify that for some architectures that qmake generates different makefiles. See line 29 to 33. e.g. armhf doesn't use the gold but the bfd linker

**Comment by magmastonealex** [comment]
Well, I was able to build for `armhf` without cross compiling!

The trick was using `gold` as the linker, not normal `ld`. It needs far less ram when linking large c++ projects

To do so, I just added `binutils-gold` to the makedepends, and passed `CONFIG+="use_gold_linker"`, and `WEBENGINE_CONFIG+="use_system_ffmpeg use_gold_linker"` to `qmake`.

7 hours later, and I have a working qtwebengine for armhf! It loads webpages fine. It's a little slow, but I believe that's because of the lack of hardware accelleration on my device.

![webengine](https://user-images.githubusercontent.com/2691799/34157976-bc236c5e-e491-11e7-9b8f-4828b8a8d76e.jpg)

**Comment by z3ntu** [comment]
Interesting. Because Debian actually explicitely disables the ld.gold linker for armhf... But good job! Now only aarch64 is missing ^^

**Comment by ollieparanoid** [comment]
Amazing!
> Well, I was able to build for `armhf` without cross compiling!

Do you mean, that you compiled it on an arm device? Or that you did not use the "native" cross-compilation method I was trying to get working for `qt5-qtwebengine`? I'm asking, because even when not using that "native" method, we're still cross-compiling with `pmbootstrap`.

More information: https://wiki.postmarketos.org/wiki/Build_internals#Cross-compile_types

**Comment by magmastonealex** [comment]
Sorry, yes, it's using the "distcc" type as it originally was, and my package was built with `pmbootstrap`.

I meant that I didn't need the "native" method for cross compiling.

**Comment by ollieparanoid** [comment]
Cool! Sounds like you made all edits in an APKBUILD already - if that is the case and you have time, could you make a PR (and possibly enable "armhf" for all luna packages, where it is disabled because of this issue)?

**Comment by ollieparanoid** [comment]
I have good news: the `aarch64` version doesn't stop at `LINK gn` anymore with the `gold` linker patch we have. I'm doing a full compile that will not land in the binary repo first, and when that ran through we can finally set `arch="all"` for all these packages again, and compile these packages for the repo. 👍

**Comment by ollieparanoid** [comment]
It turns out, the `LINK gn` hanging does not appear every time. I have built the entire package for `aarch64` as a test, as I wrote it above, and it went through. Then I changed the relevant APKBUILDs to  `arch="all"` again and started building it for the repo.

The second time it hanged again. I've stopped it and started building a third time, and it also hung. But when stopped and started another time, it finally worked **and there's a binary aarch64 package in the repo now!** It seems like once it is through compiling/linking `gn`, it's safe.

So this means the bug is still there, but it does not appear every time and at least with some effort it is possible to compile the package. We could look into this more, but I doubt that it is much use at this point - there's already the next version of `qt` out, to which Alpine will migrate at some point in the future, and maybe the bug is fixed there already. My plan is to keep this status for now and look into it more when this becomes a real blocker.
